### PR TITLE
"System.Func overload-is-better": checking if anything breaks, or not (in which case, we'd like a test)

### PR DESCRIPTION
**Pull request by smoothdeveloper**
Related to #11534 and prompted by excellent suggestion from @baronfel to apply the warning wave protocol (as was done with XML docs warnings) to tackle issues around "System.Func overload is better".

This PR is, for now, just checking if we have tests that would fail due to this feature of the compiler, and potentially adding one if not.

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
Mmh, unless I miss something, this being green may indicate we basically have a hole in the tests in terms of asserting. 

The thing I could be missing is that all the tests are run with `--compiling-fslib` stuff, which I doubt is the case.

**Comment by smoothdeveloper** [comment]
So, it seems in F# compiler codebase and tests, beside the test I've added in this PR, there is nothing which checks for "func-is-better".

This doesn't make me super confident in the test suite having nearly enough checks on the nitty gritty of overload resolution.

I'll try to hack a bit on having the warning, if that proves simple enough, it could turn this PR into fixing #11534 and prepare the ground for removing the support for "func-is-better" in a later major version.

@dsyme, would you mind sharing a bit here or in #11534 some of the historical reasons for "func-is-better" in overload resolution? and if you are favorable with a warning wave and eventual deprecation of that special casing?

On my end, for sake of explicitness and principle of least surprise, treating all delegate types the same, I'd favor taking this out in next major release (could even do this under preview guards soon), and later, work on addressing issues such as highlighted in https://github.com/dotnet/fsharp/issues/2503#issuecomment-389835185.

**Comment by smoothdeveloper** [comment]
This is green and ready for a first round of review.

**Comment by Happypig375** [comment]
Bump @cartermp

File: src/fsharp/MethodCalls.fs [hunk 1]
```diff
@@ -508,6 +508,29 @@ type CalledMeth<'T>
     /// The argument analysis for each set of curried arguments
     member x.ArgSets = argSets
 
+    member x.GetCallerArgAt index =
+      // Arrays are better...
+      let mutable currentIndex = 0
```
[the remark this answers is not in the document]
Comment (smoothdeveloper) [reply src/fsharp/MethodCalls.fs:513]: This code is problematic, I wrote it before surveying some more my assumptions (in #11544, which shows that my assumptions are just wrong).

Looking for guidance on how we want that to work to get this PR considered for merge.

Also eager to hear if there is a point or no, to move from `list list` representation of `SynValInfo`, I know @baronfel was facing slightly related question recently:

> anyone have any insight into why a named tuple-parameter is erased in the TAST?  eg
let myFun (f: (int * int * int)) = ()
in the TAST is represented as an FSharpMemberOrFunctionOrValue, whose CurriedParameterArgs is a list< list< FSharpParameter > > of length 1, whose inner list is of length 3. The three inner parameters are unnamed, but as a result of this form I lose information (as an IDE user) about the name/documentation of the parameter.

which shows that the representation as we have it, from the parser AST down to the typed AST is a bit arcane to the people not used to it.

File: tests/fsharp/typecheck/overloads/neg_at_somepoint_in_the_future_func-is-better-or-not.bsl [hunk 2]
```diff
@@ -0,0 +1,16 @@
+
+neg_at_somepoint_in_the_future_func-is-better-or-not.fsx(14,1,14,17): typecheck error FS0193: The method 'M' has overload with 'System.Func' delegate in the signature, to fix the warning, you need to specify the type explicitly.
```
[the remark this answers is not in the document]
Comment (smoothdeveloper) [reply tests/fsharp/typecheck/overloads/neg_at_somepoint_in_the_future_func-is-better-or-not.bsl:2]: Specifying the argument name may be problematic in circumstances similar to what @baronfel was dealing with, patterns and "anonymous" parameters.

If there are places we already prettify `Expr` to display to the end user, please let me know and I'll try to experiment with it.

Probably several error messages could benefit from having this kind of "expanding the expression to explain the error in actual case".

**Comment by dsyme** [comment]
@smoothdeveloper Could you include the XLF file updates as well please (see DEVGUIDE.md)

**Comment by KevinRansom** [comment]
@smoothdeveloper 
Thanks for proposing this contribution, however, it seems to have gotten a little stale.  Tests fail, there are conflicts, the last comment was fifteen months ago.

Please feel free to re-open if you wish to continue to work on this PR.

Kevin

**Comment by smoothdeveloper** [comment]
@KevinRansom, no issue, we could scope it for https://github.com/dotnet/fsharp/issues/15901, I think the implementation for it didn't give me too much trouble, so I could give it on top of the then-to-be-current implementation.

**Pull request closed without merging**